api/opensource/mfx_dispatch/linux/mfxloader.cpp

```diff
@@ -0,0 +1,583 @@
+  }
+
+  std::vector<std::string> libs;
+
+  if (MFX_IMPL_BASETYPE(par.Implementation) == MFX_IMPL_AUTO ||
+      MFX_IMPL_BASETYPE(par.Implementation) == MFX_IMPL_AUTO_ANY) {
+    libs.emplace_back(LIBMFXHW);
+    libs.emplace_back(MFX_MODULES_DIR "/" LIBMFXHW);
+    libs.emplace_back(LIBMFXSW);
+    libs.emplace_back(MFX_MODULES_DIR "/" LIBMFXSW);
+  } else if (par.Implementation & MFX_IMPL_HARDWARE ||
+             par.Implementation & MFX_IMPL_HARDWARE_ANY) {
+    libs.emplace_back(LIBMFXHW);
+    libs.emplace_back(MFX_MODULES_DIR "/" LIBMFXHW);
+  } else if (par.Implementation & MFX_IMPL_SOFTWARE) {
+    libs.emplace_back(LIBMFXSW);
+    libs.emplace_back(MFX_MODULES_DIR "/" LIBMFXSW);
+  } else {
+    return MFX_ERR_UNSUPPORTED;
+  }
+
+  mfxStatus mfx_res = MFX_ERR_UNSUPPORTED;
+
+  for (auto lib: libs) {
+    std::shared_ptr<void> hdl = make_dlopen(lib.c_str(), RTLD_LOCAL|RTLD_NOW);
+    if (hdl) {
+      do {
+        /* Loading functions table */
+        bool wrong_version = false;
+        for (int i = 0; i < eFunctionsNum; ++i) {
+          assert(i == g_mfxFuncTable[i].id);
+          m_table[i] = dlsym(hdl.get(), g_mfxFuncTable[i].name);
+          if (!m_table[i] && ((par.Version <= g_mfxFuncTable[i].version) ||
+                (g_mfxFuncTable[i].version <= mfxVersion(VERSION(1, 14))))) {
+            // this version of dispatcher requires MFXInitEx which appeared
+            // in Media SDK API 1.14
+            wrong_version = true;
+            break;
+          }
+        }
+        if (wrong_version) {
+          mfx_res = MFX_ERR_UNSUPPORTED;
+          break;
+        }
+
+        /* Initializing loaded library */
+        mfx_res = ((decltype(MFXInitEx)*)m_table[eMFXInitEx])(par, &m_session);
+        if (MFX_ERR_NONE != mfx_res) {
+          break;
+        }
+
+        // Below we just get some data and double check that we got what we have expected
+        // to get. Some of these checks are done inside mediasdk init function
+        mfx_res = ((decltype(MFXQueryVersion)*)m_table[eMFXQueryVersion])(m_session, &m_version);
+        if (MFX_ERR_NONE != mfx_res) {
+          break;
+        }
+
+        if (m_version < par.Version) {
+          mfx_res = MFX_ERR_UNSUPPORTED;
+          break;
+        }
+
+        mfx_res = ((decltype(MFXQueryIMPL)*)m_table[eMFXQueryIMPL])(m_session, &m_implementation);
+        if (MFX_ERR_NONE != mfx_res) {
+          mfx_res = MFX_ERR_UNSUPPORTED;
+          break;
+        }
+      } while(false);
+
+      if (MFX_ERR_NONE == mfx_res) {
+        m_dlh = std::move(hdl);
+        break;
+      } else {
+        Close();
+      }
+    }
+  }
+
+  return mfx_res;
+}
+
+mfxStatus LoaderCtx::Close()
+{
+  auto proc = (decltype(MFXClose)*)m_table[eMFXClose];
+  mfxStatus mfx_res = (proc)? (*proc)(m_session): MFX_ERR_NONE;
+
+  m_implementation = {};
+  m_version = {};
+  m_session = nullptr;
+  std::fill_n(m_table, sizeof(m_table)/sizeof(m_table[0]), nullptr);
```

Comment:
I suggest to use c++11 power instead :)

std::fill(std::begin(m_table), std::end(m_table), nullptr)